lib/Mojo/Redis/Protocol.pm

```diff
@@ -0,0 +1,240 @@
+package Mojo::Redis::Protocol;
+use Mojo::Base -base;
+
+use Carp qw(confess croak);
+use Encode 'find_encoding';
+
+use constant DEBUG => $ENV{MOJO_REDIS_DEBUG} || 0;
+
+our %ENCODING;    # Internal usage
+
+has on_message => undef;
+
+sub encode {
+  my ($self, @stack) = @_;
+
+  my $str = '';
+  while (@stack) {
+    my $message = shift @stack;
+    my $data    = $message->{data};
+    my $type    = $message->{type};
+
+    if ($type eq '+' or $type eq '-' or $type eq ':') {
+      $str .= "$type$data\r\n";
+    }
+    elsif ($type eq '$') {
+      $str .= defined $data ? '$' . length($data) . "\r\n$data\r\n" : "\$-1\r\n";
+    }
+    elsif ($type eq '*') {
+      if (defined $data) {
+        $str .= '*' . scalar(@$data) . "\r\n";
+        unshift @stack, @$data;
+      }
+      else {
+        $str .= "*-1\r\n";
+      }
+    }
+    else {
+      confess "[Mojo::Redis::Protocol] Unknown message type $type";
+    }
+  }
+
+  return $self->{encoder} ? $self->{encoder}->encode($str, 0) : $str;
+}
+
+sub encoding {
+  my $self = shift;
+  return $self->{encoding} unless @_;
+  $self->{encoding} = shift // '';
+  $self->{encoder} = $self->{encoding} ? ($ENCODING{$self->{encoding}} ||= find_encoding($self->{encoding})) : undef;
+  Carp::confess("Unknown encoding '$self->{encoding}'") if $self->{encoding} and !$self->{encoder};
+  return $self;
+}
+
+sub new {
+  my $self = shift->SUPER::new(@_);
+  $self->encoding($self->{encoding});
+  $self->{buf} = '';
+  return $self;
+}
+
+sub parse {
+  my $self = shift;
+  my $curr = $self->{curr} ||= {level => 0};
+  my $buf  = \$self->{buf};
+  my $encoder;
+
+  $$buf .= shift;
+
+CHUNK:
+  while (length $$buf) {
+    do { local $_ = $$buf; s!\r\n!\\r\\n!g; warn "[Mojo::Redis::Protocol] parse($_)\n" } if DEBUG > 1;
+
+    # Look for initial type
+    if (!$curr->{type}) {
+      $curr->{type} = substr $$buf, 0, 1, '';
+      _parse_debug(start => $curr) if DEBUG > 1;
+    }
+
+    # Simple Strings, Errors, Integers
+    if ($curr->{type} eq '+' or $curr->{type} eq '-' or $curr->{type} eq ':') {
+      $curr->{len} //= index $$buf, "\r\n";
+      return $curr->{len} = undef if $curr->{len} < 0;    # Wait for more data
+
+      $curr->{data} = substr $$buf, 0, $curr->{len}, '';
+      substr $$buf, 0, 2, '';                             # Remove \r\n after data
+
+      # Force into correct data type
+      if ($curr->{type} eq ':') {
+        $curr->{data} += 0;
```

Comment:
Even integers should be decoded, to allow for non-ascii encodings

Reply:
How does an non-ascii integer look?

Reply:
The integer is ascii but in a non-ascii encoding it may be encoded to a different byte (or bytes, like UTF-16)

Reply:
I guess this is moot anyway, since integer type needs to be readable by redis.

Reply:
I'm not going to do anything about this, since I don't understand how an integer can be anything else than just numbers.